feat(coderd): add `coder_app` usage stats

mafredri opened the pull request:
Updates #8658

mafredri committed:
feat(coderd): track `coder_app` usage

Updates #8658

mtojek commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,196 @@
+func (r *StatsDBReporter) Report(ctx context.Context, stats []StatsReport) error {
+	err := r.Database.InTx(func(tx database.Store) error {
+		for _, stat := range stats {
+			err := tx.InsertWorkspaceAppStats(ctx, database.InsertWorkspaceAppStatsParams{
```
I remember than Cian refactored agent stats into batches, so probably you should adjust this part as well.

mafredri replied:
Do you mean using sql COPY instead of INSERT? That could have an impact but if that wasn't what you meant, these are already batched by the collector so I don't think additional batching here will be necessary.

mtojek replied:
Something similar to https://github.com/coder/coder/commit/9fb18f3ae5c0024474439cf9de6795a2d3e7a125  , so that an array of stats is unnested within a single query.

mafredri replied:
Sure, I can implement that 👍🏻 (albeit a simpler version will do here).

mtojek replied:
This can be done later. I just wanted to raise that we identified a similar issue in the past, and decided to fix it.

mafredri replied:
I started working on this, but I identified a problem: it doesn't seem possible to define values that are nullable.

```sql
-- name: InsertWorkspaceAppStats :exec
INSERT INTO
	workspace_app_stats (
		id,
		user_id,
		workspace_id,
		agent_id,
		access_method,
		slug_or_port,
		session_id,
		session_started_at,
		session_ended_at
	)
SELECT
	unnest(@id::uuid[]) AS id,
	unnest(@user_id::uuid[]) AS user_id,
	unnest(@workspace_id::uuid[]) AS workspace_id,
	unnest(@agent_id::uuid[]) AS agent_id,
	unnest(@access_method::text[]) AS access_method,
	unnest(@slug_or_port::text[]) AS slug_or_port,
	unnest(@session_id::uuid[]) AS session_id,
	unnest(@session_started_at::timestamptz[]) AS session_started_at,
	unnest(@session_ended_at::nulltimestamptz[]) AS session_ended_at
ON CONFLICT
	(agent_id, session_id)
DO
	UPDATE SET
		-- Only session end can be updated.
		session_ended_at = EXCLUDED.session_ended_at
	WHERE
		workspace_app_stats.agent_id = EXCLUDED.agent_id
		AND workspace_app_stats.session_id = EXCLUDED.session_id;
```

I tried using `sqlc.narg` and also defining a new type (`nulltimestamptz`) in `sqlc.yaml`, but the former didn't work and the latter produces `interface{}`:

```yaml
      - db_type: "nulltimestamptz"
        nullable: true
        engine: "postgresql"
        go_type:
          import: "database/sql"
          type: "NullTime"
```

Since I'm refactoring the format, though, I might get away with not using nullable columns. We'll see.

mafredri replied:
Actually, I got this `nulltimestamptz` to work... somehow. Interesting. Not sure why it worked after another try.

mafredri replied:
As mentioned, this will still change, but: [`0bb672a` (#9001)](https://github.com/coder/coder/pull/9001/commits/0bb672ac5295d54d25c50af468305c585032fed9)

mtojek commented on enterprise/wsproxy/appstatsreporter.go:
```diff
@@ -0,0 +1,21 @@
+
+var _ workspaceapps.StatsReporter = (*appStatsReporter)(nil)
+
+type appStatsReporter struct {
```
do we expect more structures matching `StatsReporter`?

mtojek replied:
Yes, remove the interface then.

mafredri replied:
Huh? There are two implementations of stats reporter, one in `workspaceapps` (the DB one) and this one (HTTP POST one).

mtojek replied:
I missed the `StatsDBReporter`. I guess that you can mark it that it implements `StatsReport`.

~~Out of curiosity: why are they sharing the same interface? I see only `Report(..., stats []StatsReport)` function, and I'm not sure if it is worth introducing coupling between.~~

Oh I see, this is a workaround to collect data from "wsproxy".

mtojek commented on coderd/database/queries/workspaceappstats.sql:
```diff
@@ -0,0 +1,24 @@
+-- name: InsertWorkspaceAppStats :exec
+INSERT INTO
```
Can you scope this PR for collecting `app_stats` and then focus on the Insights endpoint in a follow-up? Otherwise, this PR will become too large and hard to review. As an extra benefit, we will start collecting stats earlier :)

mafredri replied:
Yup, that's the plan 👍

mafredri committed:
amend pr comments

mafredri committed:
refactor stats with rollups, add tests

mafredri commented:
I've refactored the stats collection to perform rollups on short-lived requests and added full test coverage for the `StatsCollector`.

Default rollup window is 1 minute, we can extend it if we want to reduce rows further, but I don't think it'll be necessary.

What's left is:
- Add tests for `coderd` and `wsproxy` to ensure app usage produces stats (?)
- Fix TODO about `wsproxy` auth
- Check correctness of `dbauthz` implementation

For now, the data looks like this:

```sql
SELECT *, session_ended_at - session_started_at AS session_length FROM workspace_app_stats ORDER BY session_started_at;
```
| id |user\_id |workspace\_id |agent\_id |access\_method |slug\_or\_port |session\_id |session\_started\_at |session\_ended\_at |requests |session\_length |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 24 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |path |code\-server |e250b40e\-537e\-4d89\-a45c\-852f3f0ce018 |2023\-08\-10 22:03:00\+00 |2023\-08\-10 22:04:00\+00 |9 |00:01:00 |
| 25 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |path |code\-server |16c08a01\-6827\-45ac\-959b\-fcd72b05c4bd |2023\-08\-10 22:04:00\+00 |2023\-08\-10 22:05:00\+00 |20 |00:01:00 |
| 31 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |path |code\-server |fe340b03\-46b6\-48f8\-87e6\-64b5baf708b9 |2023\-08\-10 22:04:09.204568\+00 |2023\-08\-10 22:16:20.918296\+00 |1 |00:12:11.713728 |
| 32 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |path |code\-server |291de74a\-4aee\-4ee1\-b02a\-d8cceddd56ef |2023\-08\-10 22:04:10.452269\+00 |2023\-08\-10 22:16:20.921671\+00 |1 |00:12:10.469402 |
| 30 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |terminal | |4f9c687b\-f732\-4c91\-92ac\-2cc459e6a7b2 |2023\-08\-10 22:05:00\+00 |2023\-08\-10 22:06:00\+00 |2 |00:01:00 |
| 38 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |terminal | |bd7d51df\-eb75\-4642\-9033\-3102e8ea0665 |2023\-08\-10 22:05:15.74551\+00 |2023\-08\-10 22:07:42.616165\+00 |1 |00:02:26.870655 |
| 39 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |terminal | |23e21c72\-28de\-4f12\-90e7\-7a23709599dd |2023\-08\-10 22:05:23.250785\+00 |2023\-08\-10 22:07:46.848367\+00 |1 |00:02:23.597582 |
| 40 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |terminal | |32d92f3c\-07d8\-4ed0\-9373\-d741e70cba4a |2023\-08\-10 22:05:26.718312\+00 |2023\-08\-10 22:07:44.201889\+00 |1 |00:02:17.483577 |
| 41 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |terminal | |07d531c3\-1d9b\-4eac\-b152\-2b88e2c7302a |2023\-08\-10 22:05:26.733824\+00 |2023\-08\-10 22:07:46.044149\+00 |1 |00:02:19.310325 |
| 73 |5419d403\-d7a1\-4000\-a9b1\-3bab3c9c8993 |6c7b8c1c\-aeaa\-4409\-9448\-2f46390938b6 |1dacd2af\-0b50\-4fd4\-bd99\-7443e1fe06e7 |path |code\-server |3d47059c\-50d3\-4afd\-8976\-69e836082bf0 |2023\-08\-10 22:08:00\+00 |2023\-08\-10 22:09:00\+00 |40 |00:01:00 |

mafredri committed:
add close

mafredri committed:
fix wsproxy endpoint auth todo

mafredri committed:
add backlog for re-reporting in case of failure

mafredri committed:
update dbauthz

mafredri committed:
use dbauthz.AsSystemRestricted for collector flush

mafredri committed:
add stats collection test to apptest

mafredri committed:
Merge branch 'main' into mafredri/feat-add-app-usage-to-template-insights

mafredri committed:
fix migrations

mafredri committed:
update plumbing to fix wsproxy tests

mafredri committed:
test the stat output in apptest

mafredri committed:
fix issues

mafredri committed:
fix id in dbfake

mafredri committed:
fix gen

mafredri committed:
Merge branch 'main' into mafredri/feat-add-app-usage-to-template-insights

mafredri committed:
add fixture

deansheather commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+
+		// Ensure we don't hold up this request for too long. Add a few
+		// seconds to prevent very short intervals from causing a timeout.
+		ctx, cancel := context.WithTimeout(context.Background(), sc.opts.ReportInterval+5*time.Second)
```
I would just use a fixed timeout like 15s or something, and then call `ticker.Reset` after so the interval is the duration between the end of one flush and the beginning of another flush.

deansheather commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+	rolledUp bool // Indicates if this report has been rolled up.
+}
+
+func newStatsReportFromSignedToken(token SignedToken) StatsReport {
```
If you wanted to you could add `session_id` to the SignedToken to make this easier to corroborate sessions.

mafredri replied:
Wasn't sure if I put it on either SignedToken or SignedToken.Request, will it end up somewhere? Logged or in the DB? I'll defer this to another PR if it makes sense to add.

mafredri replied:
@deansheather does this match what you had in mind? [`357eec2` (#9001)](https://github.com/coder/coder/pull/9001/commits/357eec2b293fba1654ea8589d50297c903c71c8c)

mafredri replied:
Noticed the signed tokens are also created via `Issue`, so the SessionID did not propagate.

I'm afraid my understanding of the signed token lifecycle is too limited to confidently make this change. I added this [`a23297b` (#9001)](https://github.com/coder/coder/pull/9001/commits/a23297b104804c1a3badbe661326bd8df02bf90c), but as a result the session ID will be signed too. Is that something we want or should I add it after the signature is made?

deansheather replied:
I was thinking you should put it in `db.go` where the token is originally created and signed.

deansheather replied:
Yeah I think it's fine if it's included in the signed payload. Means it persists between requests for up to a minute right?

deansheather replied:
If you don't need/want that feel free to dismiss this though

mafredri replied:
IIRC there is a certain token re-use, will there be one token created per request or will the same one (and thus, sessionID) be re-used across multiple requests?

If it's the latter, it will conflict with the rollup logic I believe.

mafredri replied:
Chatted with Dean on Discord and we decided to not try to corroborate requests/tokens for now. Since SignedTokens live for one minute and are re-used, re-using that ID would conflict with rollups of short and long-lived connections.

johnstcn commented on coderd/database/migrations/000150_workspace_app_stats.up.sql:
Note before merging: double-check for a duplicate migration!

mafredri replied:
I don't think there's any newer migration on main? But yea, I will merge in main and fix if needed before merge. 👍🏻

johnstcn commented on coderd/database/dbfake/dbfake.go:
```diff
@@ -4336,6 +4337,48 @@ func (q *FakeQuerier) InsertWorkspaceApp(_ context.Context, arg database.InsertW
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	nextID := int64(1)
+	if len(q.workspaceAppStats) > 0 {
+		nextID = q.workspaceAppStats[len(q.workspaceAppStats)-1].ID + 1
+	}
```
Why not just keep track of last insert ID in `FakeQuerier` and increment on insert? It's unfortunate btw that we have to emulate this behaviour.

mafredri replied:
Agreed it's unfortunate. The main reason I went this approach is we don't have a lot of precedent. I guess license inserts is the only one in fakedb using a last insert ID.

I'll make the change though since you asked about it. 👍🏻

mafredri commented on coderd/database/migrations/000150_workspace_app_stats.up.sql:
```diff
@@ -0,0 +1,32 @@
+CREATE TABLE workspace_app_stats (
+    id BIGSERIAL PRIMARY KEY,
+    user_id uuid NOT NULL REFERENCES users (id),
+	workspace_id uuid NOT NULL REFERENCES workspaces (id),
+    agent_id uuid NOT NULL REFERENCES workspace_agents (id),
+	access_method text NOT NULL,
+	slug_or_port text NOT NULL,
+	session_id uuid NOT NULL,
+    session_started_at timestamptz NOT NULL,
+    session_ended_at timestamptz NOT NULL,
+	requests integer NOT NULL,
+
+	-- Set a unique constraint to allow upserting the session_ended_at
+	-- and requests fields without risk of collisions.
+	UNIQUE(user_id, agent_id, session_id)
```
Good catch, will fix 👍🏻

mafredri commented on coderd/database/sqlc.yaml:
```diff
@@ -27,6 +27,13 @@ overrides:
           type: "TemplateACL"
+      # Type alias to allow batch inserts of timestamptz[] arrays containing
+      # nullable values.
+      - db_type: "nulltimestamptz"
```
Not anymore, I'll remove it 👍🏻

mafredri commented on coderd/workspaceapps/proxy.go:
```diff
@@ -586,7 +592,13 @@ func (s *Server) proxyWorkspaceApp(rw http.ResponseWriter, r *http.Request, appT
+	report := newStatsReportFromSignedToken(appToken)
+	s.collectStats(report)
+
 	proxy.ServeHTTP(rw, r)
```
Thanks for catching that!

johnstcn commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+// Report writes the given StatsReports to the database.
+func (r *StatsDBReporter) Report(ctx context.Context, stats []StatsReport) error {
+	err := r.db.InTx(func(tx database.Store) error {
+		var batch database.InsertWorkspaceAppStatsParams
```
Instead of inserting in a tight loop, would we be better off doing a bulk insert?

mafredri replied:
I think that's what we're doing (see batchSize), or did you mean something else?

johnstcn commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+
+	// Options for tests.
+	Flush <-chan chan<- struct{}
+	Now   func() time.Time
```
Suggestion: allow setting a `chan time.Time` for use in unit tests. This will allow more fine-grained control over time.

mafredri replied:
I think the now func is more flexible than a chan. We can even implement the chan with it, if we want: `Now: func() time.Time { <-testTimeCh }`.

johnstcn replied:
Fair :+1:

mafredri commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+				rollupChanged = true
+				continue
+			}
+			if stat.SessionEndedAt.IsZero() && sc.opts.Now().Sub(stat.SessionStartedAt) <= sc.opts.RollupWindow {
```
Solid suggestion! 💪🏻

mafredri committed:
use 15s timeout and timer reset

mafredri committed:
pass now to rollup

mafredri committed:
track last insert id in fakedb

mafredri committed:
fix migration indentation

mafredri committed:
remove sqlc type alias

mafredri committed:
defer collect stats

mafredri committed:
remove stale comment

mafredri commented on coderd/workspaceapps/stats.go:
```diff
@@ -0,0 +1,378 @@
+	err := r.db.InTx(func(tx database.Store) error {
+		var batch database.InsertWorkspaceAppStatsParams
+		for _, stat := range stats {
+			batch.UserID = append(batch.UserID, stat.UserID)
+			batch.WorkspaceID = append(batch.WorkspaceID, stat.WorkspaceID)
+			batch.AgentID = append(batch.AgentID, stat.AgentID)
+			batch.AccessMethod = append(batch.AccessMethod, string(stat.AccessMethod))
+			batch.SlugOrPort = append(batch.SlugOrPort, stat.SlugOrPort)
+			batch.SessionID = append(batch.SessionID, stat.SessionID)
+			batch.SessionStartedAt = append(batch.SessionStartedAt, stat.SessionStartedAt)
+			batch.SessionEndedAt = append(batch.SessionEndedAt, stat.SessionEndedAt)
+			batch.Requests = append(batch.Requests, int32(stat.Requests))
```
Should maybe test, but I don't think this will escape to the heap, so mostly just stack allocations which are very cheap and don't load the gc. I can make the change nonetheless but trying to be more efficient here is more error prone when fields are added or removed.

mafredri replied:
Pushed the change 👍🏻

mafredri committed:
preallocate batch slices

mafredri committed:
Merge branch 'main' into mafredri/feat-add-app-usage-to-template-insights

mafredri committed:
lower log level to debug for start/stop
